### PR TITLE
[Fixbug][Perf] Qwen3-omni: code predictor with re-prefill + SDPA and eliminate decode hot-path CPU round-trips

### PR DESCRIPTION
**Pull request by LJH-LBJ**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Resolves: https://github.com/vllm-project/vllm-omni/issues/1830
This PR fixes two critical accuracy bugs in the Qwen3-Omni code predictor that caused audio quality degradation, and re-applies the re-prefill + SDPA optimization from #1758 with correctness guarantees.

Performance analysis and detailed design remain the same as described in https://github.com/vllm-project/vllm-omni/pull/1758 — **this PR does not regress performance**.

### Root Cause 1: `_proj_buf` persistent buffer cross-request pollution

PR #1758 introduced a persistent `self._proj_buf` tensor (allocated once in `__init__`) to accumulate embeddings across the autoregressive loop. When multiple concurrent requests share the same model instance, request A's embedding data written into `_proj_buf` gets silently overwritten by request B, causing subsequent AR steps to read corrupted history. This is the primary cause of intermittent accuracy failures under concurrent load.

**Fix:** Allocate `proj_buf` locally in every `forward()` call:

```python
# Before (buggy): persistent buffer shared across requests
self._proj_buf = torch.zeros(max_batch, max_seq, hidden_size, ...)  # in __init__

# After (fixed): per-call allocation, no aliasing
def forward(self, ...):
    proj_buf = torch.zeros(bsz, max_seq, self._hidden_size, dtype=dtype, device=device)
```
### Root Cause 2: `summed_embeddings` 3D + `text_step` 2D broadcast error

In `talker_mtp()`, `summed_embeddings` has shape `[B, S, H]` (3D) while `text_step` has shape `[B*S, H]` (2D). When added directly, PyTorch performs silent broadcasting — for `B > 1` the result is completely wrong (`[B, 1, H] + [B, H]` → `[B, B, H]`).

**Fix:** Flatten `summed_embeddings` to 2D before addition:

```python
# Before (buggy): 3D + 2D silent broadcast
inputs_embeds = summed_embeddings.clone()
inputs_embeds = (inputs_embeds + text_step).reshape(...)

# After (fixed): uniform 2D addition
inputs_embeds = summed_embeddings.reshape(-1, self.talker_config.text_config.hidden_size)
inputs_embeds = (inputs_embeds + text_step).reshape(...)
```

## Test Plan
Accurancy Test twice
```
pytest -s -v   tests/e2e/online_serving/test_qwen3_omni_expansion.py   -m "advanced_model" --run-level "advanced_model"
```

benchmark
```
vllm serve /workspace/models/Qwen3-Omni-30B-A3B-Instruct --omni --port 46354 --stage-configs-path ./vllm_omni/model_executor/stage_configs/qwen3_omni_moe_async_chunk.yaml
```

## Test Result
image

image

benchmark
| Metric | Before | After |
|--------|--------|-------|
| TTFP (ms) | 13000.91 | 8851.15 |
| E2E (ms) | 86184.13 | 83231.48 |
| RTF | 0.65 | 0.58 |
| TTFT (ms) | 6877.48 | 5329.86 |

```
Before:
============ Serving Benchmark Result ============							
Successful requests:                     10        							
Failed requests:                         0         							
Maximum request concurrency:             10        							
Benchmark duration (s):                  131.49    							
Request throughput (req/s):              0.08      							
Peak concurrent requests:                10.00     							
----------------End-to-end Latency----------------							
Mean E2EL (ms):                          86184.13  							
Median E2EL (ms):                        95894.25  							
P99 E2EL (ms):                           130903.62 							
================== Text Result ===================							
Total input tokens:                      25000     							
Total generated tokens:                  5199      							
Output token throughput (tok/s):         39.54     							
Peak output token throughput (tok/s):    306.00    							
Peak concurrent requests:                10.00     							
Total Token throughput (tok/s):          229.67    							
---------------Time to First Token----------------							
Mean TTFT (ms):                          6877.48   							
Median TTFT (ms):                        7451.55   							
P99 TTFT (ms):                           7454.69   							
-----Time per Output Token (excl. 1st token)------							
Mean TPOT (ms):                          101.27    							
Median TPOT (ms):                        81.24     							
P99 TPOT (ms):                           240.85    							
---------------Inter-token Latency----------------							
Mean ITL (ms):                           53.44     							
Median ITL (ms):                         0.01      							
P99 ITL (ms):                            1925.43   							
================== Audio Result ==================							
Total audio duration generated(s):       1669.25   							
Total audio frames generated:            40061925  							
Audio throughput(audio duration/s):      12.69     							
---------------Time to First Packet---------------							
Mean AUDIO_TTFP (ms):                    13000.91  							
Median AUDIO_TTFP (ms):                  13079.55  							
P99 AUDIO_TTFP (ms):                     13637.70  							
-----------------Real Time Factor-----------------							
Mean AUDIO_RTF:                          0.65      							
Median AUDIO_RTF:                        0.57      							
P99 AUDIO_RTF:                           1.12      							
==================================================							


After:
============ Serving Benchmark Result ============							
Successful requests:                     10        							
Failed requests:                         0         							
Maximum request concurrency:             10        							
Benchmark duration (s):                  124.16    							
Request throughput (req/s):              0.08      							
Peak concurrent requests:                10.00     							
----------------End-to-end Latency----------------							
Mean E2EL (ms):                          83231.48  							
Median E2EL (ms):                        85536.25  							
P99 E2EL (ms):                           122818.30 							
================== Text Result ===================							
Total input tokens:                      25000     							
Total generated tokens:                  5172      							
Output token throughput (tok/s):         41.66     							
Peak output token throughput (tok/s):    304.00    							
Peak concurrent requests:                10.00     							
Total Token throughput (tok/s):          243.02    							
---------------Time to First Token----------------							
Mean TTFT (ms):                          5329.86   							
Median TTFT (ms):                        5882.53   							
P99 TTFT (ms):                           5884.73   							
-----Time per Output Token (excl. 1st token)------							
Mean TPOT (ms):                          104.86    							
Median TPOT (ms):                        92.06     							
P99 TPOT (ms):                           198.48    							
---------------Inter-token Latency----------------							
Mean ITL (ms):                           51.24     							
Median ITL (ms):                         0.01      							
P99 ITL (ms):                            2015.94   							
================== Audio Result ==================							
Total audio duration generated(s):       1618.88   							
Total audio frames generated:            38853060  							
Audio throughput(audio duration/s):      13.04     							
---------------Time to First Packet---------------							
Mean AUDIO_TTFP (ms):                    8851.15   							
Median AUDIO_TTFP (ms):                  8848.69   							
P99 AUDIO_TTFP (ms):                     9442.12   							
-----------------Real Time Factor-----------------							
Mean AUDIO_RTF:                          0.58      							
Median AUDIO_RTF:                        0.58      							
P99 AUDIO_RTF:                           0.83      							
==================================================												

```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
any performance w/o the feature here?

**Comment by LJH-LBJ** [comment]
> any performance w/o the feature here?

I didn’t make any changes to the main feature branch. The bug fix only addressed a tensor dimension mismatch and added a torch.zeros in one place; nothing else was modified. I don’t think it has any impact on performance.

**Comment by amy-why-3459** [comment]
Please provide a performance data comparison before and after this PR.

**Comment by amy-why-3459** [comment]
Please provide accuracy test results under multiple concurrent scenarios. Please add an accuracy test under multiple concurrent scenarios.

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 1]
```diff
@@ -159,6 +159,12 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
             # suppress tokens by setting their probability to ~1e-9 (finite very small)
             self.suppressed_tokens = self._get_talker_suppressed_tokens()
             self.requires_raw_input_tokens = True
+            # Keys that should stay on GPU in model_intermediate_buffer to avoid CPU↔GPU round-trips
+            self.gpu_resident_buffer_keys: set[str] = {
+                "last_talker_hidden",
+                "trailing_text_hidden",
+                "tts_pad_embed_projected",
```
[the remark this answers is not in the document]
Comment (LJH-LBJ) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:166]: code_predictor_codes will not be readed from gpu_resident_buffer_keys  . no need to add to gpu_resident_buffer_keys

**Comment by LJH-LBJ** [comment]
> Please provide accuracy test results under multiple concurrent scenarios. Please add an accuracy test under multiple concurrent scenarios.

multiple concurrent scenarios already exists in test_qwen3_omni_expansion.py
 you can check the request_num in the test case

**Comment by hsliuustc0106** [comment]
> > Please provide accuracy test results under multiple concurrent scenarios. Please add an accuracy test under multiple concurrent scenarios.
> 
> multiple concurrent scenarios already exists in test_qwen3_omni_expansion.py you can check the request_num in the test case

I suggest to add the perf w/o this PR which is necessary to show the gain or regression.

**Comment by LJH-LBJ** [comment]
> > > Please provide accuracy test results under multiple concurrent scenarios. Please add an accuracy test under multiple concurrent scenarios.
> > 
> > 
> > multiple concurrent scenarios already exists in test_qwen3_omni_expansion.py you can check the request_num in the test case
> 
> I suggest to add the perf w/o this PR which is necessary to show the gain or regression.

OK， I will test it

**Comment by amy-why-3459** [comment]
@Sy0307 PTAL

**Comment by LJH-LBJ** [comment]
> any performance w/o the feature here?

@hsliuustc0106 @amy-why-3459 Already posted, I test with [PR1982](https://github.com/vllm-project/vllm-omni/pull/1982)

**Comment by Sy0307** [comment]
LGTM. Related issue is https://github.com/vllm-project/vllm-omni/pull/2019. It will wait for this PR being merged.

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni_moe_code_predictor_mtp.py [hunk 2]
```diff
@@ -88,120 +83,55 @@ def __init__(
...
-                f"All attention backends failed. Last error: {last_error}. "
-                "Please install flash-attn, or ensure PyTorch's scaled_dot_product_attention is available."
-            )
-        attn_output = attn_output.reshape(*(hidden_states.shape[:-1]), -1).contiguous()
+        # [B, heads, seq, head_dim]
+        q = q.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        k = k.view(bsz, seq_len, self.num_kv_heads, self.head_dim).transpose(1, 2)
+        v = v.view(bsz, seq_len, self.num_kv_heads, self.head_dim).transpose(1, 2)
+
+        attn_out = F.scaled_dot_product_attention(
```
Comment (ZeldaHuang) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni_moe_code_predictor_mtp.py:114]: can we use this?
https://github.com/vllm-project/vllm/blob/9040151fe1899aba6e2934364fb4c5edfcb5e29c/vllm/v1/attention/ops/vit_attn_wrappers.py#L190

Comment (LJH-LBJ) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni_moe_code_predictor_mtp.py:114]: `apply_sdpa` from `vllm/v1/attention/ops/vit_attn_wrappers.py` **cannot** be used here — it omits `is_causal=True`.

The code predictor runs autoregressive re-prefill: each AR step forwards a growing sequence (length 2 → `num_code_groups + 1`). Position *i* must only attend to positions ≤ *i*; otherwise the yet-to-be-predicted slots (initialized to zero in `proj_buf`) leak into the attention output and corrupt the logits. `is_causal=True` enforces the lower-triangular mask that prevents this.

`apply_sdpa` is designed for ViT-style encoders where every token attends to every other token (bidirectional). Dropping the causal mask here would break codec generation quality.
